File: workshop/example/python/parens.py
from parser import Consecutive, character, lazy, epsilon, just, succeed

class Tree:
    pass

class Nil(Tree):
    def __eq__(self, other):
        return isinstance(other, Nil)
    
    def __repr__(self):
        return 'Nil()'

class Bin(Tree):
    def __init__(self, first, second):
        self.first = first
        self.second = second
    
    def __eq__(self, other):
        return isinstance(other, Bin) and self.first == other.first and self.second == other.second

    def __repr__(self):
        return f'Bin({self.first},{self.second})'

def foldparens(combine, initial):
    parser = Consecutive([
        character('('),
        lazy(lambda : parser),
        character(')'),
        lazy(lambda : parser)
    ]).map(combine).orElse(succeed(initial))
    return parser


def parens():
    return foldparens(lambda result: Bin(result[1], result[3]), Nil())

def nesting():
    return foldparens(lambda result: max(1 + result[1], result[3]), 0)

if __name__ == '__main__':
    assert Nil() == Nil()
    assert Bin(Nil(), Bin(Nil(), Nil())) == Bin(Nil(), Bin(Nil(), Nil()))
    assert just(parens())('()(())') == [(Bin(Nil(),Bin(Bin(Nil(),Nil()),Nil())), '')]

    depth = just(nesting())
    assert depth('') == [(0, '')]
    assert depth('()') == [(1, '')]
    assert depth('()(())') == [(2, '')]

File: workshop/example/node/parens.js
const assert = require('assert').strict
const Parser = require('./parser')
const Util = require('./util')

class Nil {
    constructor() {}

    representation() {
        return "Nil()"
    }

    equals(other) {
        return other instanceof Nil
    }
}

class Bin {
    constructor(left, right) {
        this.left = left
        this.right = right
    }

    representation() {
        return "Bin(" + 
            this.left.representation() + 
            "," + 
            this.right.representation() +
            ")"
    }

    equals(other) {
        if (other instanceof Bin) {
            return this.left.equals(other.left) && this.right.equals(other.right)
        }
        return false
    }
}

function foldparens(combine, initial){
    const parser = Parser.or(
        Parser.map(Util.consecutive([
            Parser.character('('),
            Parser.lazy(function(){ return parser }),
            Parser.character(')'),
            Parser.lazy(function(){ return parser})
        ]), combine),
        Parser.succeed(initial)
    )
    return parser
}

const parens = foldparens(
    function(result){return new Bin(result[1],result[3])},
    new Nil()
)

const nesting = foldparens(
    function(result){return Math.max(1 + result[1], result[3])},
    0
)

assert.ok(new Nil().equals(new Nil()))
assert.ok(
    new Bin(new Nil(), new Bin(new Nil(), new Nil())).equals(
    new Bin(new Nil(), new Bin(new Nil(), new Nil()))))
const result = Parser.just(parens)("()(())")
assert.ok(
    new Bin(new Nil(), new Bin(new Bin(new Nil(), new Nil()), new Nil())).equals(
    result[0][0]))
